podman/api/ssh.py

```diff
@@ -1,75 +1,300 @@
 """
-from typing import Any, Mapping, Optional, Union
-from urllib.parse import urlparse
+import collections
+import functools
+import http.client
+import logging
+import pathlib
+import random
+import socket
+import subprocess
+import time
+import urllib.parse
+from contextlib import suppress
+from typing import Optional, Union
 
-from requests.adapters import HTTPAdapter
-from requests.packages.urllib3 import HTTPConnectionPool  # pylint: disable=import-error
-from requests.packages.urllib3.connection import HTTPConnection  # pylint: disable=import-error
-from requests.packages.urllib3.util import Timeout  # pylint: disable=import-error
+import xdg.BaseDirectory
+from requests.adapters import DEFAULT_POOLBLOCK, DEFAULT_RETRIES, HTTPAdapter
 
+try:
+    import urllib3
+except ImportError:
+    import requests.packages.urllib3 as urllib3
 
-class SSHConnection(HTTPConnection):
-    """Specialization of HTTPConnection to use a ssh tunnel."""
+from .adapter_utils import _key_normalizer
 
-    def __init__(
-        self,
-        host: str,
-        timeout: Optional[Union[float, Timeout]] = None,
-    ):
-        """Instantiate connection to ssh tunnel for Podman service for HTTP client."""
-        _ = host
-        _ = timeout
-        raise NotImplementedError
+logger = logging.getLogger("podman.ssh_adapter")
+
+
+class SSHSocket(socket.socket):
+    """Specialization of socket.socket to forward a UNIX domain socket via SSH."""
+
+    def __init__(self, uri: str, identity: Optional[str] = None):
+        """Initialize SSHSocket.
+
+        Args:
+            uri: Full address of a Podman service including path to remote socket.
+            identity: path to file containing SSH key for authorization
+
+        Examples:
+            SSHSocket("http+ssh://alice@api.example:2222/run/user/1000/podman/podman.sock",
+                      "~alice/.ssh/api_ed25519"
+            )
+        """
+        super().__init__(socket.AF_UNIX, socket.SOCK_STREAM)
+        self.uri = uri
+        self.identity = identity
+        self._proc: Optional[subprocess.Popen] = None
+
+        runtime_dir = pathlib.Path(xdg.BaseDirectory.get_runtime_dir(strict=False)) / "podman"
+        runtime_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+
+        self.local_sock = runtime_dir / f"podman-forward-{random.getrandbits(80):x}.sock"
+
+    def connect(self, **kwargs):  # pylint: disable=unused-argument
+        """Returns socket for SSH tunneled UNIX domain socket.
+
+        Raises:
+            subprocess.TimeoutExpired: when SSH client fails to create local socket
+        """
+        uri = urllib.parse.urlparse(self.uri)
+
+        command = [
+            "ssh",
+            "-N",
+            "-L",
+            f"{self.local_sock}:{uri.path}",
+        ]
+
+        if self.identity is not None:
+            path = pathlib.Path(self.identity).expanduser()
+            command += ["-i", str(path)]
+
+        command += [f"ssh://{uri.netloc}"]
```

Comment:
here.  I think you need to break out the port number into a `-p` option.  Also I'm not sure about using the `ssh://` prefix...I've never used that before (maybe it works?).

Reply:
Oh okay, I see in the man page now:

> ssh connects and logs into the specified destination, which may be specified
> as either [user@]hostname or a URI of the form ssh://[user@]hostname[:port]

Hrmmmm.